sql/core/src/test/scala/org/apache/spark/sql/connector/functions/FunctionBenchmark.scala

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.functions
+
+import org.apache.spark.benchmark.Benchmark
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.connector.catalog.{Identifier, InMemoryCatalog}
+import org.apache.spark.sql.connector.catalog.functions.{BoundFunction, ScalarFunction, UnboundFunction}
+import org.apache.spark.sql.execution.benchmark.SqlBasedBenchmark
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{DataType, LongType, StructType}
+
+/**
+ * Benchmark to measure DataSourceV2 UDF performance
+ * {{{
+ *   To run this benchmark:
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class>
+ *        --jars <spark core test jar>,<spark catalyst test jar> <sql core test jar>
+ *   2. build/sbt "sql/test:runMain <this class>"
+ *   3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
+ *      Results will be written to "benchmarks/FunctionBenchmark-results.txt".
+ * }}}
+ */
+object FunctionBenchmark extends SqlBasedBenchmark {
+  val catalogName: String = "benchmark_catalog"
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
+    val N = 500L * 1000 * 1000
+    scalarBenchmark(N, resultNullable = false)
+    scalarBenchmark(N, resultNullable = true)
+  }
+
+  private def scalarBenchmark(N: Long, resultNullable: Boolean): Unit = {
+    withSQLConf(s"spark.sql.catalog.$catalogName" -> classOf[InMemoryCatalog].getName) {
```

Comment:
Just for my curiosity/education, why do we need to override the catalog here? Can't we just use the default in-memory catalog?

Reply:
Ah. It appears I'm mixing up `spark.sql.catalogImplementation` (which defaults to `org.apache.spark.sql.catalyst.catalog.InMemoryCatalog`) vs. `spark.sql.catalog` (which  defaults to `V2SessionCatalog` but you are overriding it to `org.apache.spark.sql.connector.catalog.InMemoryCatalog`). These two similarly named configurations and classes are quite confusing!

Reply:
Ah yea, it's easy to confuse the two. We can perhaps change its name if necessary. There are a few other classes with the same name between DSv1 and v2, e.g., `Table`, `AggregateFunction`,